### PR TITLE
Add code to allow testing stream-k grid multipliers.

### PR DESCRIPTION
**Pull request by AlexBrownAMD**
* Add code to allow testing stream-k grid multipliers.
* Fix race condition that appeared with large grids and small sizes.
* Add debug output to check occupancy.

### CODE REVIEW

File: Tensile/Source/lib/source/ContractionSolution.cpp [hunk 1]
```diff
@@ -35,6 +35,8 @@
 #include <cstdlib>
 #include <random>
 
+const int TENSILE_STREAMK_GRID 1
```
Comment (nakajee) [inline Tensile/Source/lib/source/ContractionSolution.cpp:38]: don't we need ; here?

File: Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -12719,6 +12719,8 @@ def partialsWriteProcedure(self, kernel, vectorWidths, elements, alpha, beta, ed
       kStr += PreLoopVmcntCaseStr
 
     # Set flag
+    kStr += inst("s_waitcnt", "lgkmcnt(0)", "wait for data store")
```
Comment (nakajee) [inline Tensile/KernelWriterAssembly.py:12722]: Shouldn't this be vmcnt?

File: Tensile/Source/lib/source/ContractionSolution.cpp (hunk 1 above)
Comment (AlexBrownAMD) [reply Tensile/Source/lib/source/ContractionSolution.cpp:38]: Yea, I had this as a define while testing, started to change it before posting but must have messed up the saved version. I've changed it back to define.

File: Tensile/KernelWriterAssembly.py (hunk 2 above)
Comment (AlexBrownAMD) [reply Tensile/KernelWriterAssembly.py:12722]: Yes it should, thanks! Posted a fix

File: Tensile/Source/lib/source/hip/HipSolutionAdapter.cpp [hunk 3]
```diff
@@ -346,6 +346,16 @@ namespace Tensile
                                        &argsSize,
                                        HIP_LAUNCH_PARAM_END};
 
+            if(m_debug)
+            {
+                int numBlocks = 0;
+                int blockSize
+                    = kernel.workGroupSize.x * kernel.workGroupSize.y * kernel.workGroupSize.z;
+                HIP_CHECK_RETURN(hipModuleOccupancyMaxActiveBlocksPerMultiprocessor(
```
Comment (nakajee) [inline Tensile/Source/lib/source/hip/HipSolutionAdapter.cpp:354]: I would like to understand this part a bit more.
- how the output look like? Could you share some example?
- Why 4th parameter(dynSharedMemPerBlk ) is 0? I do not quite understand the interface of this function.

Comment (AlexBrownAMD) [reply Tensile/Source/lib/source/hip/HipSolutionAdapter.cpp:354]: This prints the occupancy of the kernel, I was using it to check occupancy of stream-k vs non-stream-k kernel variants. Output looks like:

Occupancy = 2

dynSharedMemPerBlk is amount of dynamic LDS we plan to allocate. We don't use any dynamic LDS, we know how much shared memory is required when we generate the kernel, so this is set to 0. The function can check .amdhsa_group_segment_fixed_size or .group_segment_fixed_size in kernel header for how much fixed shared memory is required to do the occupancy calculation.

Comment (nakajee) [reply Tensile/Source/lib/source/hip/HipSolutionAdapter.cpp:354]: Thanks for your explanation. I got it.

**Comment by nakajee** [comment]
Why integration test failed?

**Review by nakajee — APPROVED** [review]
All tests passed. Looks good.

**Pull request merged**